File: 20_ValidParentheses/20_v1.java
public class Solution {
    public boolean isValid(String s) {
        HashMap mapping = new HashMap<Character, Character>();
        mapping.put(']', '[');
        mapping.put('}', '{');
        mapping.put(')', '(');
        
        Stack check = new Stack<Character>();
        
        for (int i = 0; i < s.length(); i++){
            if (mapping.get(s.charAt(i)) != null){
                if (check.isEmpty() || mapping.get(s.charAt(i)) != check.peek())
                    return false;
                check.pop();
            } else {
                check.push(s.charAt(i));
            }
        }
        
        return check.isEmpty();
    }
}

File: 20_ValidParentheses/20_v1.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        mapping = {
            ')': '(',
            ']': '[',
            '}': '{',
        }
        
        check = []
        
        for c in s:
            if c in mapping:
                if not check or mapping[c] != check[-1]:
                    return False
                check.pop()
            else:
                check.append(c)
        
        return not check
